Runtime/ObjectMap.cs

```diff
@@ -0,0 +1,184 @@
+/// .NET objects to be automatically wrapped when returned to JS, and re-wrapped as needed
+/// if the (weakly-referenced) JS wrapper has been released.
+/// </summary>
+internal static class ObjectMap
+{
+    // TODO: Consider an optimization that avoids dictionary lookups:
+    // If an exported class is declared as partial, then the generator can add a
+    // static `JSConstructor` property and an instance `JSWrapper` property to the class.
+    // (The dictionary mappings could still be used to export external/non-partial classes.)
+
+    private static readonly ConcurrentDictionary<Type, JSReference> s_classMap = new();
+    private static readonly ConcurrentDictionary<object, JSReference> s_objectMap = new();
+    private static readonly ConcurrentDictionary<Type, JSReference> s_structMap = new();
+
+    /// <summary>
+    /// Registers a class JS constructor, enabling automatic JS wrapping of instances of the class.
+    /// </summary>
+    /// <param name="constructorFunction">JS class constructor function returned from
+    /// <see cref="JSNativeApi.DefineClass"/></param>
+    /// <returns>The JS constructor.</returns>
+    internal static JSValue RegisterClass<T>(JSValue constructorFunction) where T : class
+    {
+        s_classMap.AddOrUpdate(
+            typeof(T),
+            (_) => new JSReference(constructorFunction, isWeak: false),
+            (_, _) => throw new InvalidOperationException(
+                "Class already registered for JS export: " + typeof(T)));
+        return constructorFunction;
+    }
+
+    /// <summary>
+    /// Gets a class JS constructor that was previously registered.
+    /// </summary>
+    private static JSValue GetClassConstructor<T>() where T : class
+    {
+        if (!s_classMap.TryGetValue(typeof(T), out JSReference? constructorReference))
+        {
+            throw new InvalidOperationException(
+                "Class not registered for JS export: " + typeof(T));
+        }
+
+        JSValue? constructorFunction = constructorReference!.GetValue();
+        if (!constructorFunction.HasValue)
+        {
+            // This should never happen because the reference is "strong".
+            throw new InvalidOperationException("Failed to resolve class constructor reference.");
+        }
+
+        return constructorFunction.Value;
+    }
+
+    /// <summary>
+    /// Attaches an object to a JS wrapper, and saves a weak reference to the wrapper.
+    /// </summary>
+    /// <param name="wrapper">JS object passed as the 'this' argument to the constructor callback
+    /// for <see cref="JSNativeApi.DefineClass"/>.</param>
+    /// <param name="obj">New or existing instance of the class to be wrapped.</param>
+    /// <returns>The JS wrapper.</returns>
+    internal static unsafe JSValue InitializeObjectWrapper<T>(JSValue wrapper, T obj) where T : class
+    {
+        GCHandle valueHandle = GCHandle.Alloc(obj);
+        napi_ref result;
+        napi_wrap(
```

Comment:
>napi_wrap [](http://example.com/codeflow?start=8&length=9)

Can we use or add method from/to JSNativeApi.

Reply:
Yes. I'll add a `JSNativeApi.Wrap()` overload that returns the weak reference. The current one does not.

Reply:
Updated.